Add support for OpenId strategy for authentication

israel-hdez opened the pull request:
* Add support for OpenId strategy for authentication.
  * It was just a matter of adding openid to some conditionals.
  * Renamed OpenshiftLogin class as OAuthLogin for code clarity
  * (see commits cdd8ffac07ca1a8de459e0768c4e434f40cb8fc3 and c6415eafa8c741c304f9b5869036566fd1aab048)
* Move the checkCredentials to AuthenticationController.tsx
  * This was mainly for a more centralized authentication flow.
  * Also, to prevent showing the login form for OpenShift/OpenID while the login cycle is finalizing.
  * (see commit 3defea26de510bb7ae34c39126d4f5861e8a55c1)

Part of kiali/kiali#2056

israel-hdez commented:
Promoting this PR. IMO, it looks ready.

jshaughn commented on src/app/AuthenticationController.tsx:
```diff
@@ -57,6 +61,36 @@ class AuthenticationController extends React.Component<AuthenticationControllerP
+        authenticationConfig.strategy === AuthStrategy.openshift ||
+        authenticationConfig.strategy === AuthStrategy.openid
+      ) {
+        const pattern = /[#&](access_token|id_token)=/;
```
I learned something new, regex literals, neato!

jshaughn commented on src/app/AuthenticationController.tsx:
```diff
@@ -86,6 +120,16 @@ class AuthenticationController extends React.Component<AuthenticationControllerP
+        !this.state.isPostLoginError &&
+        (authenticationConfig.strategy === AuthStrategy.openshift ||
+          authenticationConfig.strategy === AuthStrategy.openid)
+      ) {
```
This multi-line conditional keeps getting repeated, perhaps for readability just add a simple private function isAuthStrategyOath() or something like that.

israel-hdez replied:
I added the helper function. Not private, because this condition is on several files.

jmazzitelli commented:
> We do want to deprecate or remove the login strategy right?

Why would we want to do that? This is the default mechanism used by upstream Istio users, I don't think we want to remove that. For simple installs, I thin the login strategy is fine to use. Maybe we want to change the *default* strategy (perhaps make the default "token", instead of "login")?

jshaughn commented:
> Why would we want to do that? This is the default mechanism used by upstream Istio users

I think we discussed this [in Edgar's proposal doc](https://docs.google.com/document/d/1Zc4QXEUm4wfa5u8NQv_8RdzQokwsfrOR7F8os_uGch4).  The plan was to remove LDAP and Login, and make the default "Token", like you mentioned above.

israel-hdez commented:
I rebased.
@jshaughn I added three commits after your  review. Not sure if you would like to check those three additional commits as @mtho11 has already reviewed.

@mtho11 I fixed the linting issue.

jmazzitelli commented:
@israel-hdez a very trivial conflict needs to be fixed - prettier must have removed a comma

jmazzitelli commented:
@israel-hdez I ran a test. I cannot login when using the Dex setup you provided. Running that script is successful (BTW: I'm going to check that script into our repo - probably in the kiali/kiali hack/ directory, but maybe under the operator's molecule/ directory if I can write a molecule test for this).
But when I try to login as "admin@example.com" with password of "password", the Kiali login screen tells me it is invalid:
![image](https://user-images.githubusercontent.com/2029470/83343607-a7c4af80-a2ca-11ea-9e23-5659f551a9ae.png)

This is different behavior if I actually type the wrong password. If I type "foo" for password in the OpenID login screen, I see this:

![image](https://user-images.githubusercontent.com/2029470/83343621-cfb41300-a2ca-11ea-8704-17e14813303f.png)

israel-hdez commented:
@jmazzitelli OK, I see the missing part on that minikube-openid gist script is, most likely, assign privileges.
In the script, I forgot to create the `rolebindings` so that the `admin@example.com` users has access to a namespace.

You could either:
* Give it cluster-admin privileges: `minikube kubectl -- create clusterrolebinding openid-admin-binding --clusterrole=cluster-admin --user="admin@example.com"`
* Give it cluster-wide privs to Kiali: `minikube kubectl -- create clusterrolebinding openid-kiali-admin-binding --clusterrole=kiali --user="admin@example.com"`
* Give it Kiali privs to one namespace: `minikube kubectl -- create rolebinding openid-kiali-bookinfo-binding --clusterrole=kiali --user="admin@example.com" --namespace=bookinfo`

Once the user has some privileges, you should be able to properly login in Kiali.

jmazzitelli commented:
@israel-hdez I tried that but get the same results. I installed bookinfo, and kiali (built with your PRs).

I then gave the user permission to see bookinfo:

```
$ kubectl get clusterrole kiali
NAME    CREATED AT
kiali   2020-06-01T17:25:45Z
$ kubectl get namespace bookinfo
NAME       STATUS   AGE
bookinfo   Active   103m
$ minikube -p openid kubectl -- create rolebinding openid-rolebinding-bookinfo --clusterrole=kiali --user=admin@example.com --namespace=bookinfo
rolebinding.rbac.authorization.k8s.io/openid-rolebinding-bookinfo created
```

The auth section of the Kiali CR:
```
  auth:
    openid:
      client_id: kiali-app
      insecure_skip_verify_tls: true
      issuer_uri: https://192-168-99-100.nip.io:32000
      username_claim: email
    strategy: openid
```
I get back to the Kiali login screen with the red message, "Unauthorized. The provided credentials are not valid to access Kiali. Please check your credentials and try again."

Logs:

```
$ kubectl -n istio-system logs kiali-7467699f45-vv5c9 
I0601 18:02:50.020499       1 kiali.go:66] Kiali: Version: v1.19.0-SNAPSHOT, Commit: 3aee3cba64b08d8d029c5f921baa04a27ee14fa6
I0601 18:02:50.020552       1 kiali.go:67] DEBUG: Kiali: Command line: [/opt/kiali/kiali -config /kiali-configuration/config.yaml -v 4]
I0601 18:02:50.020558       1 config.go:527] DEBUG: Reading YAML config from [/kiali-configuration/config.yaml]
I0601 18:02:50.021060       1 kiali.go:205] Using authentication strategy [openid]
I0601 18:02:50.021099       1 kiali.go:87] Kiali: Console version: 1.19.0-local-b2a37d3ea5e41a783cd4d0554876f024882c0d32
I0601 18:02:50.021105       1 kiali.go:284] Updating base URL in index.html with [/kiali]
I0601 18:02:50.021132       1 kiali.go:297] DEBUG: Base URL has been updated to [<base href="/kiali/"]
I0601 18:02:50.021444       1 kiali.go:265] Generating env.js from config
I0601 18:02:50.021481       1 kiali.go:270] DEBUG: The content of /opt/kiali/console/env.js will be:
window.WEB_ROOT='/kiali';
I0601 18:02:50.024533       1 server.go:67] Server endpoint will start at [:20001/kiali]
I0601 18:02:50.024551       1 server.go:68] Server endpoint will serve static content from [/opt/kiali/console]
I0601 18:02:50.024568       1 metrics_server.go:18] Starting Metrics Server on [:9090]
I0601 18:04:10.248246       1 client.go:244] DEBUG: Rest perf config QPS: 175.000000 Burst: 200
I0601 18:04:10.250616       1 cache.go:126] Kiali Cache is active for namespaces [.*]
I0601 18:04:10.250987       1 client.go:244] DEBUG: Rest perf config QPS: 175.000000 Burst: 200
I0601 18:04:18.371927       1 client.go:244] DEBUG: Rest perf config QPS: 175.000000 Burst: 200
```

I confirmed that SHA (3aee3cba64b08d8d029c5f921baa04a27ee14fa6) is for the commit where I merged in your PR - so I know this is the code from the PR.

israel-hdez commented:
Conflict is resolved.

israel-hdez commented:
Passing the ball to QA.
